File: Random/0009.Gas Station.py
#    Author : Mohamed Yousef 
#    Date   : 2023-01-07
import sys,math,bisect,collections,itertools,heapq
from collections import defaultdict,deque
sys.setrecursionlimit(50000) #To increase Recursion depth in py

def solve(gas,cost):
    n=len(cost)
    profit=0
    big=0
    start=-1
    for i in range(n):
        profit+=(gas[i]-cost[i])    
        big+=(gas[i]-cost[i])    
        if big<0:
            big=0
            start=i+1
    print(start)
    return -1 if profit<0 else start
print(solve([1,2,3,4,5], [3,4,5,1,2]))

File: Random/0009.Gas Station.js
/*
    Author : Mohamed Yousef 
    Date   : 2023-01-07
*/
let solve=function(gas,cost){
    let start=0;
    let now=0
    profit=0
    n=gas.length;
    for(let i=0;i<n;i++){
        profit+=(gas[i]-cost[i])
        now+=(gas[i]-cost[i])
        if(now<0){
            now=0
            start=i+1
        }
    }
    // console.log(start);
    return profit<0 ? -1: start;
}
console.log(solve([1,2,3,4,5],[3,4,5,1,2]));